core/src/main/java/io/smallrye/openapi/runtime/io/ModelIO.java

```diff
@@ -206,8 +221,168 @@ public T read(AnnotationValue annotation) {
+            if (desiredType.type == DataType.Type.MAP) {
+                Map<String, Object> result = new HashMap<>();
+                O object = jsonIO().asObject(node);
+                for (Entry<String, V> entry : jsonIO().properties(object)) {
+                    result.put(entry.getKey(), readJson(entry.getValue(), desiredType.content));
+                }
+                return result;
+            } else if (desiredType.type == DataType.Type.OBJECT) {
+                if (desiredType.clazz == Object.class) {
+                    return jsonIO().fromJson(node);
+                } else {
+                    return readValue(node, desiredType.clazz);
+                }
+            } else {
+                // Log node dropped?
+                return null;
+            }
+        } else if (jsonIO().isArray(node)) {
+            if (desiredType.type == DataType.Type.LIST) {
+                List<Object> result = new ArrayList<>();
+                A array = jsonIO().asArray(node);
+                for (V element : jsonIO().entries(array)) {
+                    result.add(readJson(element, desiredType.content));
+                }
+                return result;
+            } else if (desiredType.clazz == Object.class) {
+                return jsonIO().fromJson(node);
+            } else {
+                // Log node dropped?
+                return null;
+            }
+        } else if (desiredType.type == DataType.Type.OBJECT) {
+            if (desiredType.clazz == Object.class) {
+                return jsonIO().fromJson(node);
+            } else {
+                return readValue(node, desiredType.clazz);
+            }
+        } else {
+            return jsonIO().fromJson(node);
+        }
+    }
+
+    /**
+     * Convert JSON value node to an object when we have a desired type
+     * <p>
+     * The JSON value will be converted to the desired type if possible or returned as its native type if not.
+     *
+     * @param node the JSON node
+     * @param desiredType the type that we want to be returned
+     * @return an object which represents the JSON node, which may or may not be of the desired type
+     */
+    @SuppressWarnings("unchecked")
+    protected Object readValue(V node, Class<?> desiredType) {
+        if (desiredType == Schema.class) {
+            return schemaIO().readValue(node);
+        }
+
+        // Handles string, number and boolean types
+        Object result = jsonIO().fromJson(node, desiredType);
+        if (result != null) {
+            return result;
+        }
+
+        if (Enum.class.isAssignableFrom(desiredType)) {
+            result = enumValue(node, desiredType.asSubclass(Enum.class));
+            if (result != null) {
+                return result;
+            }
+        }
+
+        if (Constructible.class.isAssignableFrom(desiredType)) {
+            return readObject((Class<? extends Constructible>) desiredType, (O) node);
+        }
```

Comment:
I was able to reproduce it also with this input.

```yaml
---
openapi: 3.1.0
info:
  title: My Title
  version: "1.0.0"
  contact: null
paths: {}
```

Maybe something like this to avoid it?

```java
if (Constructible.class.isAssignableFrom(desiredType)) {
    if (jsonIO().isObject(node)) {
        return readObject((Class<? extends Constructible>) desiredType, (O) node);
    } else if (jsonIO().isNull(node)) { // new `isNull` method
        return null;
    } else {
        // Log that value type cannot be read as a desiredType.getName(); 
        return null;
    }
}
```

Reply:
That would work.

My initial thought was
```java
        if (Constructible.class.isAssignableFrom(desiredType) && jsonIO().isObject(node)) {
            return readObject((Class<? extends Constructible>) desiredType, (O) node);
        }
```
and allow unexpected types to fall through to `jsonIO().fromJson(node)`, but that causes unexpected types to end up in the model.

Dropping the unexpected type on read would match the old IO objects, though `SchemaIO` allowed unexpected types in and did a type check in its getter methods. It looks like with this PR that `SchemaIO` no longer does that which I'd have thought would have caused tests to fail so I might be missing something there.

Reply:
I would also expect a failure if invalid types were allowed in before. Ultimately they should be written back out as long as the version is 3.1.0. Let me know what you find on that. I suppose it should be OK unless they've set an alternate schema dialect, in which case everything should pass through as-is.